### PR TITLE
Intel switchdev

### PR DESCRIPTION
**Pull request by zeeke**
Ice driver supports creating VFs after the eSwitcMode is
set to switchdev. This is different than the preferred way
for `mlx5` driver.

Add specific functions to configure eSwitchMode and SriovNumVFs
for each driver.

Add a basic test to loop over all available
devices and test if setting `eSwitchMode: switchdev` works.

### CODE REVIEW

**Review by ykulazhenkov — DISMISSED** [review]
LGTM

**Comment by Eoghan1232** [comment]
I will do a test of this

**Comment by Eoghan1232** [comment]
worked like a charm @zeeke  ;) 

let me give the PR a review now then we should be good to merge

**Review by Eoghan1232 — APPROVED** [review]
Changes LGTM!

once this is merged, we should also update the support HW list.
I will do this

File: pkg/host/internal/sriov/sriov.go [hunk 1]
```diff
@@ -986,22 +986,30 @@ func (s *sriov) createVFs(iface *sriovnetworkv1.Interface) error {
...
+	pfDriverName, err := s.dputilsLib.GetDriverName(pciAddr)
+	if err != nil {
 		return err
 	}
-	if err := s.SetNicSriovMode(pciAddr, eswitchMode); err != nil {
-		err = fmt.Errorf("failed to switch NIC to SRIOV %s mode: %v", eswitchMode, err)
-		log.Log.Error(err, "setEswitchMode(): failed to set mode", "device", pciAddr, "mode", eswitchMode)
-		return err
+
+	type setEswitchModeAndNumVFsFn func(string, string, int) error
```
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:995]: nit: can we move this one outside the function?

File: pkg/host/internal/sriov/sriov.go [hunk 2]
```diff
@@ -986,22 +986,30 @@ func (s *sriov) createVFs(iface *sriovnetworkv1.Interface) error {
...
+
+	var setEswitchModeAndNumVFsByDriverName map[string]setEswitchModeAndNumVFsFn = map[string]setEswitchModeAndNumVFsFn{
```
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:997]: nit: can we do something like

```
setEswitchModeAndNumVFsByDriverName map[string]setEswitchModeAndNumVFsFn := map[string]setEswitchModeAndNumVFsFn{...}
```

File: pkg/host/internal/sriov/sriov.go [hunk 3]
```diff
@@ -986,22 +986,30 @@ func (s *sriov) createVFs(iface *sriovnetworkv1.Interface) error {
 	return s.setEswitchModeAndNumVFs(iface.PciAddress, expectedEswitchMode, iface.NumVfs)
 }
 
-func (s *sriov) setEswitchMode(pciAddr, eswitchMode string) error {
-	log.Log.V(2).Info("setEswitchMode(): set eswitch mode", "device", pciAddr, "mode", eswitchMode)
-	if err := s.unbindAllVFsOnPF(pciAddr); err != nil {
-		log.Log.Error(err, "setEswitchMode(): failed to unbind VFs", "device", pciAddr, "mode", eswitchMode)
+func (s *sriov) setEswitchModeAndNumVFs(pciAddr string, desiredEswitchMode string, numVFs int) error {
+	pfDriverName, err := s.dputilsLib.GetDriverName(pciAddr)
```
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:990]: nit: can we please add a debug log level about the PCI address and the driver name?

File: pkg/host/internal/sriov/sriov.go [hunk 4]
```diff
@@ -986,22 +986,30 @@ func (s *sriov) createVFs(iface *sriovnetworkv1.Interface) error {
...
+	var setEswitchModeAndNumVFsByDriverName map[string]setEswitchModeAndNumVFsFn = map[string]setEswitchModeAndNumVFsFn{
+		"ice":       s.setEswitchModeAndNumVFsIce,
+		"mlx5_core": s.setEswitchModeAndNumVFsMlx,
 	}
-	return nil
+
+	fn, ok := setEswitchModeAndNumVFsByDriverName[pfDriverName]
+	if !ok {
+		// Fallback to mlx5 driver
+		fn = s.setEswitchModeAndNumVFsMlx
```
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:1005]: can you please add a debug message that we didn't find the driver in the supported list so we fallback

File: pkg/host/internal/sriov/sriov.go [hunk 5]
```diff
@@ -986,22 +986,30 @@ func (s *sriov) createVFs(iface *sriovnetworkv1.Interface) error {
...
+	}
+
+	return fn(pciAddr, desiredEswitchMode, numVFs)
 }
 
-func (s *sriov) setEswitchModeAndNumVFs(pciAddr string, desiredEswitchMode string, numVFs int) error {
-	log.Log.V(2).Info("setEswitchModeAndNumVFs(): configure VFs for device",
+func (s *sriov) setEswitchModeAndNumVFsMlx(pciAddr string, desiredEswitchMode string, numVFs int) error {
```
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:1011]: please add a function comment to explain the process of how we configure switchDev virtual functions for this case

File: pkg/host/internal/sriov/sriov.go [hunk 6]
```diff
@@ -1011,13 +1019,46 @@ func (s *sriov) setEswitchModeAndNumVFs(pciAddr string, desiredEswitchMode strin
...
 	}
 
 	if desiredEswitchMode == sriovnetworkv1.ESwithModeSwitchDev {
 		return s.setEswitchMode(pciAddr, sriovnetworkv1.ESwithModeSwitchDev)
 	}
+
+	return nil
+}
+
+func (s *sriov) setEswitchModeAndNumVFsIce(pciAddr string, desiredEswitchMode string, numVFs int) error {
```
Comment (SchSeba) [inline pkg/host/internal/sriov/sriov.go:1034]: please add a function comment to explain the process of how we configure switchDev virtual functions for this case

File: test/conformance/tests/test_switchdev.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("[sriov] Switchdev", Ordered, func() {
+
+	BeforeAll(func() {
+		if cluster.VirtualCluster() {
+			Skip("IGB driver does not support VF statistics")
```
Comment (SchSeba) [inline test/conformance/tests/test_switchdev.go:25]: nit: please update the skip message here :)

File: test/conformance/tests/test_switchdev.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+		sriovInfos, err := cluster.DiscoverSriov(clients, operatorNamespace)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(len(sriovInfos.Nodes)).ToNot(BeZero())
+
+		testNode, interfaces, err := sriovInfos.FindSriovDevicesAndNode()
+		Expect(err).ToNot(HaveOccurred())
+
+		By(fmt.Sprintf("Testing on node %s, %d devices found", testNode, len(interfaces)))
+
+		mainDevice := findMainSriovDevice(getConfigDaemonPod(testNode), interfaces)
```
Comment (SchSeba) [inline test/conformance/tests/test_switchdev.go:47]: you can use `findUnusedSriovDevices` to have the list of non main device :)

File: pkg/host/internal/sriov/sriov.go (hunk 1 above)
Comment (zeeke) [reply pkg/host/internal/sriov/sriov.go:995]: ok, moving. though that type is not supposed to be used out of this function

File: pkg/host/internal/sriov/sriov.go (hunk 4 above)
Comment (zeeke) [reply pkg/host/internal/sriov/sriov.go:1005]: good point!

**Comment by adrianchiris** [comment]
Please hold off with merging this one id like to go over this pr by end of week

File: pkg/host/internal/sriov/sriov.go [hunk 9]
```diff
@@ -1029,6 +1047,40 @@ func (s *sriov) setEswitchModeAndNumVFs(pciAddr string, desiredEswitchMode strin
 	return nil
 }
 
+// setEswitchModeAndNumVFsIce configures PF eSwitch and sriov_numvfs in the following order:
+// a. set eSwitchMode to `switchdev` if requested
```
Comment (Eoghan1232) [inline pkg/host/internal/sriov/sriov.go:1051]: we need to reset vf's to 0 before setting switchdev mode, is this handled?

File: pkg/host/internal/sriov/sriov.go (hunk 5 above)
Comment (adrianchiris) [reply pkg/host/internal/sriov/sriov.go:1024]: nit: between b. and c. we also unbind all VFs on PF.
can you add this to the comment ?

note: rebind happens after part of `configSriovVFDevices`

also what about `configureHWOptionsForSwitchdev` ? is it the same for intel ice ?

**Review by adrianchiris — COMMENTED** [review]
overall lgtm, left one small comment

File: pkg/host/internal/sriov/sriov.go (hunk 9 above)
Comment (Eoghan1232) [reply pkg/host/internal/sriov/sriov.go]: I believe so, otherwise driver will throw an error I believe. 
I can double check this tomorrow to confirm

Comment (Eoghan1232) [reply pkg/host/internal/sriov/sriov.go]: I can see in some docs that we must have 0 VF's before switching to switchdev mode.

File: pkg/host/internal/sriov/sriov.go (hunk 5 above)
Comment (Eoghan1232) [reply pkg/host/internal/sriov/sriov.go:1025]: I don't think we need to unbind driver, since we need to reset back to 0 vf's anyways

Comment (zeeke) [reply pkg/host/internal/sriov/sriov.go:1025]: Got it. Then better move the `unbindAllVFsOnPF(...)` call outside of `setEswitchMode(...)`

Comment (adrianchiris) [reply pkg/host/internal/sriov/sriov.go:1025]: > Regarding configureHWOptionsForSwitchdev, current implementation is about setting flow_steering_mode=smfs, which is Mellanox only (IIUC).

this also enables hw-tc-offload via ethtool.

Comment (zeeke) [reply pkg/host/internal/sriov/sriov.go:1025]: > > Regarding configureHWOptionsForSwitchdev, current implementation is about setting flow_steering_mode=smfs, which is Mellanox only (IIUC).
> 
> this also enables hw-tc-offload via ethtool.

right! I miss that. I checked and it should not be a problem for the ice driver:

from a local test: 
```
2024-08-29T08:21:49.27421628Z   LEVEL(-2)       sriov/sriov.go:365      EnableHwTcOffload(): enable offloading  {"device": "eno12409"}
2024-08-29T08:21:49.274526639Z  LEVEL(-2)       sriov/sriov.go:365      EnableHwTcOffload(): already enabled    {"device": "eno12409"}
```

and `flow_steering_mode` setting is guarded: if it's not supported, it doesn't hurt.

**Review by SchSeba — APPROVED** [review]
LGTM

nice work!

**Review by e0ne — APPROVED** [review]

**Comment by zeeke** [comment]
Got approval from multiple sides, merging.

**Pull request merged**